python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -720,14 +719,56 @@ def _compute_sys_path(self):
     # Internal methods
 
     @sgtk.LogManager.log_timing
-    def _cache_actions(self, data, config_data):
+    def _cache_actions(self, data, config_data, cached_contents_hash=None, async=False):
         """
         Triggers the caching or recaching of engine commands.
 
         :param dict data: The data passed down from the wss client.
         :param dict config_data: A dictionary that contains, at a minimum,
             "lookup_hash", "contents_hash", "descriptor", and "entity" keys.
+        :param str cached_contents_hash: If given, represents the currently
+            cached contents hash for the configuration. If this is the same as
+            a newly-generated contents hash built prior to caching, then the
+            caching process will stop and exit without doing any additional
+            work. This represents the situation where we've been asked to
+            re-cache actions, but we then prove that the existing cached data
+            is still valid.
+        :param bool async: Indicates whether the caching routine should be
+            run asynchronously.
         """
+        descriptor = config_data["descriptor"]
+        lookup_hash = config_data["lookup_hash"]
+
+        if async:
+            now = time.time()
+            if lookup_hash in self.CACHE_VALIDATED:
+                time_since = now - self.CACHE_VALIDATED[lookup_hash]
+                if time_since < self.CACHE_VALIDATION_INTERVAL:
+                    logger.debug(
+                        "Recaching of data for %s has already been initiated. "
+                        "This thread will exit without triggering a recache of "
+                        "actions for this entry.", lookup_hash
+                    )
+                    return
+                else:
+                    self.CACHE_VALIDATED[lookup_hash] = now
```

Comment:
I see some redundant logic here with setting the cached time. I'll clean it up before merging.